File: backend/app/portfolio/schemas/portfolio.py
"""Pydantic v2 schemas for Phase 11.0: Portfolio Intelligence Foundation."""

from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.portfolio.constants import (
    DEFAULT_LOOKBACK_DAYS,
    PORTFOLIO_VERSION,
    BenchmarkTier,
    PortfolioStatus,
    TrendDirection,
)


class PortfolioQueryParams(BaseModel):
    """Query parameters for portfolio trend filtering."""
    lookback_days: int = Field(default=DEFAULT_LOOKBACK_DAYS, ge=1, le=365)


class WorkspacePortfolioEntry(BaseModel):
    """Summary of an individual workspace within a portfolio snapshot."""
    workspace_id: UUID
    workspace_name: str
    health_score: float
    rank: int
    total_ranked: int
    percentile: float
    percentile_rank: float
    benchmark_tier: BenchmarkTier
    benchmark_available: bool
    trend_direction: TrendDirection
    finding_count: int
    critical_finding_count: int
    recommendation_count: int
    last_snapshot_at: Optional[datetime] = None
    snapshot_age_seconds: Optional[float] = None


class PortfolioSummaryResponse(BaseModel):
    """Executive portfolio summary aggregating all workspaces in an organization."""
    organization_id: UUID
    portfolio_status: PortfolioStatus
    workspace_count: int
    analyzed_workspace_count: int
    portfolio_health_score: Optional[float] = None
    average_health_score: Optional[float] = None
    median_health_score: Optional[float] = None
    benchmark_available: bool
    best_workspace: Optional[WorkspacePortfolioEntry] = None
    worst_workspace: Optional[WorkspacePortfolioEntry] = None
    workspaces: List[WorkspacePortfolioEntry] = Field(default_factory=list)
    message: Optional[str] = None
    portfolio_version: str = PORTFOLIO_VERSION
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class WorkspaceBenchmarkResponse(BaseModel):
    """Detailed benchmark standing for a single workspace."""
    organization_id: UUID
    workspace_id: UUID
    workspace_name: str
    health_score: float
    rank: int
    total_ranked: int
    percentile: float
    percentile_rank: float
    benchmark_tier: BenchmarkTier
    benchmark_available: bool
    kpi_score: Optional[float] = None
    finding_count: int = 0
    critical_finding_count: int = 0
    recommendation_count: int = 0
    forecast_confidence: Optional[float] = None
    benchmark_date: datetime


class PortfolioRankingResponse(BaseModel):
    """Complete ranked leaderboard of workspaces across an organization."""
    organization_id: UUID
    rankings: List[WorkspaceBenchmarkResponse] = Field(default_factory=list)
    total_workspaces: int = 0
    benchmark_available: bool = False
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class PortfolioHealthResponse(BaseModel):
    """Portfolio health breakdown, distribution tiers, and critical attention list."""
    organization_id: UUID
    portfolio_status: PortfolioStatus
    portfolio_health_score: Optional[float] = None
    average_health_score: Optional[float] = None
    median_health_score: Optional[float] = None
    benchmark_available: bool = False
    health_distribution: Dict[str, int] = Field(default_factory=dict)
    critical_workspaces: List[WorkspacePortfolioEntry] = Field(default_factory=list)
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class PortfolioTrendPoint(BaseModel):
    """A historical snapshot data point in a portfolio health time series."""
    date: datetime
    average_health_score: Optional[float] = None
    workspace_count: int = 0
    portfolio_status: PortfolioStatus = PortfolioStatus.INSUFFICIENT_DATA


class PortfolioTrendResponse(BaseModel):
    """Historical health trend for an organization's portfolio over a lookback window."""
    organization_id: UUID
    trend_direction: TrendDirection
    lookback_days: int
    trend_points: List[PortfolioTrendPoint] = Field(default_factory=list)
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class PortfolioComparisonResponse(BaseModel):
    """Direct side-by-side benchmark comparison between two workspaces."""
    organization_id: UUID
    workspace_a: WorkspaceBenchmarkResponse
    workspace_b: WorkspaceBenchmarkResponse
    health_score_delta: float
    rank_delta: int
    generated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: frontend/src/types/portfolio.ts
/**
 * Portfolio Intelligence Suite – Typed response interfaces.
 *
 * Derived from verified OpenAPI schema at /api/v1/openapi.json:
 *   GET /api/v1/portfolio/summary          → PortfolioSummaryResponse
 *   GET /api/v1/portfolio/executive/risk   → PortfolioRiskSummary
 */

// ---------------------------------------------------------------------------
// Enum-style string unions (mirrors backend Python enums)
// ---------------------------------------------------------------------------

export type PortfolioStatus = 'HEALTHY' | 'AT_RISK' | 'CRITICAL' | 'INSUFFICIENT_DATA';
export type TrendDirection = 'UP' | 'DOWN' | 'STABLE';
export type RiskLevel = 'LOW' | 'MODERATE' | 'HIGH' | 'CRITICAL';

// ---------------------------------------------------------------------------
// WorkspacePortfolioEntry — the core "portfolio row" returned by the backend
// ---------------------------------------------------------------------------

export interface WorkspacePortfolioEntry {
  workspace_id: string;
  workspace_name: string;
  health_score: number;
  rank: number;
  total_ranked: number;
  percentile: number;
  percentile_rank: number;
  benchmark_tier: string;
  benchmark_available: boolean;
  trend_direction: TrendDirection;
  finding_count: number;
  critical_finding_count: number;
  recommendation_count: number;
  last_snapshot_at?: string | null;
  snapshot_age_seconds?: number | null;
}

// ---------------------------------------------------------------------------
// Portfolio Summary (Capital Allocation + Portfolio Rollup)
// GET /api/v1/portfolio/summary
// ---------------------------------------------------------------------------

export interface PortfolioSummaryResponse {
  organization_id: string;
  portfolio_status: PortfolioStatus;
  workspace_count: number;
  analyzed_workspace_count: number;
  portfolio_health_score?: number | null;
  average_health_score?: number | null;
  median_health_score?: number | null;
  benchmark_available: boolean;
  best_workspace?: WorkspacePortfolioEntry | null;
  worst_workspace?: WorkspacePortfolioEntry | null;
  workspaces: WorkspacePortfolioEntry[];
  message?: string | null;
  portfolio_version?: string;
  generated_at: string;
}

// ---------------------------------------------------------------------------
// Portfolio Risk Summary (Risk Concentration Radar)
// GET /api/v1/portfolio/executive/risk
// ---------------------------------------------------------------------------

export interface PortfolioRiskSummary {
  total_at_risk_workspaces: number;
  total_critical_workspaces: number;
  risk_concentration_percent: number;
  risk_level: RiskLevel;
  portfolio_size: number;
  ranked_workspace_count: number;
  generated_at: string;
}
